### PR TITLE
Added Cloning and Shortcuts for Gates to QuantumComputations

### PR DESCRIPTION
**Pull request by hillmich**
- QuantumComputations can be cloned now
- Shortcuts to add StandardOperations and Measurements
- Tests

### CODE REVIEW

File: include/QuantumComputation.hpp [hunk 1]
```diff
@@ -207,6 +228,101 @@ namespace qc {
...
+        void peres(dd::Qubit target0, dd::Qubit target1, const dd::Controls& controls) { emplace_back<StandardOperation>(getNqubits(), controls, target0, target1, qc::Peres); }
+
+        void peresdag(dd::Qubit target0, dd::Qubit target1) { emplace_back<StandardOperation>(getNqubits(), dd::Controls{}, target0, target1, qc::Peresdag); }
+        void peresdag(dd::Qubit target0, dd::Qubit target1, const dd::Control& control) { emplace_back<StandardOperation>(getNqubits(), dd::Controls{control}, target0, target1, qc::Peresdag); }
+        void peresdag(dd::Qubit target0, dd::Qubit target1, const dd::Controls& controls) { emplace_back<StandardOperation>(getNqubits(), controls, target0, target1, qc::Peresdag); }
+
+        void measure(dd::Qubit qubit, std::size_t clbit) { emplace_back<NonUnitaryOperation>(getNqubits(), qubit, clbit); }
+        void measure(const std::vector<dd::Qubit>& qubitRegister, const std::vector<std::size_t>& classicalRegister) { emplace_back<NonUnitaryOperation>(getNqubits(), qubitRegister, classicalRegister); }
+
```
Comment (burgholzer) [inline include/QuantumComputation.hpp:325]: `reset` and `barrier` instruction shortcuts are missing here

File: include/operations/ClassicControlledOperation.hpp [hunk 2]
```diff
@@ -37,6 +37,11 @@ namespace qc {
             type         = ClassicControlled;
         }
 
+        std::unique_ptr<Operation> clone() override {
+            auto op_cloned = op->clone();
+            return std::make_unique<ClassicControlledOperation>(op_cloned, controlRegister, expectedValue);
```
Comment (burgholzer) [inline include/operations/ClassicControlledOperation.hpp:42]: Cloning of a classic-controlled operation is not tested.

File: include/operations/CompoundOperation.hpp [hunk 3]
```diff
@@ -28,6 +28,16 @@ namespace qc {
...
 
+        std::unique_ptr<Operation> clone() override {
+            std::unique_ptr<CompoundOperation> cloned_ops;
+            cloned_ops->reserve(ops.size());
+
+            for (auto const& op: ops) {
+                cloned_ops->ops.emplace_back<>(op->clone());
+            }
+            return cloned_ops;
+        }
```
Comment (burgholzer) [inline include/operations/CompoundOperation.hpp]: Cloning of a compound operation is not tested.

File: include/operations/NonUnitaryOperation.hpp [hunk 4]
```diff
@@ -37,6 +37,18 @@ namespace qc {
...
+            if (getType() == qc::Measure) {
+                return std::make_unique<NonUnitaryOperation>(getNqubits(), getTargets(), getClassics());
+            } else if (getType() == qc::Snapshot) {
+                return std::make_unique<NonUnitaryOperation>(getNqubits(), getTargets(), getParameter().at(0));
+            } else if (getType() == qc::ShowProbabilities) {
+                return std::make_unique<NonUnitaryOperation>(getNqubits());
+            } else {
+                return std::make_unique<NonUnitaryOperation>(getNqubits(), getTargets(), getType());
+            }
+        }
```
Comment (burgholzer) [inline include/operations/NonUnitaryOperation.hpp:50]: Cloning of a some non-unitary operations is not tested.

File: src/operations/NonUnitaryOperation.cpp [hunk 5]
```diff
@@ -11,7 +11,9 @@ namespace qc {
     // Measurement constructor
     NonUnitaryOperation::NonUnitaryOperation(const dd::QubitCount nq, std::vector<dd::Qubit> qubitRegister, std::vector<std::size_t> classicalRegister):
         qubits(std::move(qubitRegister)), classics(std::move(classicalRegister)) {
-        assert(qubitRegister.size() == classicalRegister.size());
+        if (qubitRegister.size() != classicalRegister.size()) {
```
Comment (burgholzer) [inline src/operations/NonUnitaryOperation.cpp]: Maybe include a test for this case?

File: test/CMakeLists.txt [hunk 6]
```diff
@@ -40,7 +40,8 @@ macro(package_add_test TESTNAME)
 endmacro()
 
 # add unit tests
-package_add_test(${PROJECT_NAME}_test ${CMAKE_CURRENT_SOURCE_DIR}/algorithms/test_qft.cpp
+package_add_test(${PROJECT_NAME}_test
+				 ${CMAKE_CURRENT_SOURCE_DIR}/algorithms/test_qft.cpp
```
Comment (burgholzer) [inline test/CMakeLists.txt:44]: did the indentation change on purpose?

File: test/unittests/test_qfr_functionality.cpp [hunk 7]
```diff
@@ -749,3 +749,81 @@ TEST_F(QFRFunctionality, removeFinalMeasurementsWithOperationsInFront) {
...
+    qc.peres(0, 1, {2_pc, 3_nc});
+    qc.peresdag(0, 1);
+    qc.peresdag(0, 1, 2_pc);
+    qc.peresdag(0, 1, {2_pc, 3_nc});
+    qc.measure(0, 0);
+    qc.measure({1, 2}, {1, 2});
+
+    auto qc_cloned = qc.clone();
+    ASSERT_EQ(qc.size(), qc_cloned.size());
+}
```
Comment (burgholzer) [inline test/unittests/test_qfr_functionality.cpp]: no newline at end

File: test/unittests/test_qfr_functionality.cpp [hunk 8]
```diff
@@ -749,3 +749,81 @@ TEST_F(QFRFunctionality, removeFinalMeasurementsWithOperationsInFront) {
...
+    qc.iswap(0, 1, 2_pc);
+    qc.iswap(0, 1, {2_pc, 3_nc});
+    qc.peres(0, 1);
+    qc.peres(0, 1, 2_pc);
+    qc.peres(0, 1, {2_pc, 3_nc});
+    qc.peresdag(0, 1);
+    qc.peresdag(0, 1, 2_pc);
+    qc.peresdag(0, 1, {2_pc, 3_nc});
+    qc.measure(0, 0);
+    qc.measure({1, 2}, {1, 2});
```
Comment (burgholzer) [inline test/unittests/test_qfr_functionality.cpp:825]: this could include the missing `barrier` and `reset` shortcuts.

**Review by burgholzer — CHANGES_REQUESTED** [review]
This looks great! 👍🏼 
In my opinion it just needs 2-3 additional shortcuts and a few tests before it can be merged.

File: test/CMakeLists.txt (hunk 6 above)
Comment (hillmich) [reply test/CMakeLists.txt:44]: as discussed

**Review by burgholzer — APPROVED** [review]

**Pull request merged**